src/util/amounts.ts

```diff
@@ -0,0 +1,92 @@
+import BN from "bn.js";
+import { assert } from "chai";
+import { Fraction } from "../fraction";
+
+const MAX128 = new BN(2).pow(new BN(128)).subn(1);
+
+type SellAmount = BN;
+type BuyAmount = BN;
+
+/**
+ * Modifies the price to work with ERC20 units
+ * @param price - Amount of buy token in exchange for one sell token
+ * @param sellTokenDecimals - Number of decimals of the sell token
+ * @param buyTokenDecimals - Number of decimals of the buy token
+ * Returns Fraction representing the number of buy tokens in exchange for one unit of sell token
+ */
+export function getUnitPrice(
+  price: number,
+  sellTokenDecimals: number,
+  buyTokenDecimals: number,
+): Fraction {
+  assert.isTrue(
+    sellTokenDecimals >= 0,
+    "sell token decimals must be non-negative",
+  );
+  assert.isTrue(
+    buyTokenDecimals >= 0,
+    "buy token decimals must be non-negative",
+  );
+
+  return Fraction.fromNumber(price).mul(
+    new Fraction(
+      new BN(10).pow(new BN(buyTokenDecimals)),
+      new BN(10).pow(new BN(sellTokenDecimals)),
+    ),
+  );
+}
+
+/**
+ * Computes the amount of output token units from their price and the amount of input token units
+ * Note that the price is expressed in terms of tokens, while the amounts are in terms of token units
+ * @param price - Amount of buy token in exchange for one sell token
+ * @param sellAmount - Amount of sell token units that are exchanged at price
+ * @param sellTokenDecimals - Number of decimals of the sell token
+ * @param buyTokenDecimals - Number of decimals of the buy token
+ * Returns amount of output token units obtained
+ */
+export function getBuyAmountFromPrice(
+  price: number,
+  sellAmount: SellAmount,
+  sellTokenDecimals: number,
+  buyTokenDecimals: number,
+): BuyAmount {
+  const unitPrice = getUnitPrice(price, sellTokenDecimals, buyTokenDecimals);
+  const buyAmount = unitPrice.mul(new Fraction(sellAmount, 1));
+  return buyAmount.toBN();
```

Comment:
Typescript question! The returned value is supposed to be of type `BuyAmount` but I'm fairly confident that `fraction.toBN()` is of type `BN`. Since there is no casting to `BuyAmount`, I expect TS to just treat `BuyAmount` as a synonym for `BN`. Are there practical implications of using a custom type or is it more akin to some sort of "type comment"?

Reply:
I am probably not the best to ask here on this, I would assume it is treated as a synonym.